src/Catty/FormulaEditor/ViewController/FormulaEditorSectionViewControllerDataExtension.swift

```diff
@@ -0,0 +1,303 @@
+
+    func askForNewVariableName() {
+        Util.askUser(forVariableNameAndPerformAction: #selector(saveVariable(name:)),
+                     target: self,
+                     promptTitle: kUIFENewVarExists,
+                     promptMessage: kUIFEOtherName,
+                     minInputLength: UInt(kMinNumOfVariableNameCharacters),
+                     maxInputLength: UInt(kMaxNumOfVariableNameCharacters),
+                     isList: false,
+                     andTextField: nil,
+                     initialText: "")
+    }
+
+    @objc func saveVariable(name: String) {
+        if self.newVarIsForProject {
+            if let project = self.spriteObject?.scene.project {
+                for variable in UserDataContainer.allVariables(for: project) where variable.name == name {
+                    self.askForNewVariableName()
+                    return
+                }
+            }
+        } else {
+            if let object = self.spriteObject {
+                for variable in UserDataContainer.objectAndProjectVariables(for: object) where variable.name == name {
+                    self.askForNewVariableName()
+                    return
+                }
+            }
+        }
+
+        let userVariable = UserVariable(name: name)
+        userVariable.value = Int(0)
+
+        if self.newVarIsForProject {
+            self.spriteObject?.scene.project?.userData.add(userVariable)
+        } else {
+            self.spriteObject?.userData.add(userVariable)
+        }
+
+        self.spriteObject?.scene.project?.saveToDisk(withNotification: false)
+        self.reloadData()
+    }
+
+    @objc func saveList(name: String) {
+        if self.newVarIsForProject {
+            if let project = self.spriteObject?.scene.project {
+                for variable in UserDataContainer.allLists(for: project) where variable.name == name {
+                    self.askForNewVariableName()
+                    return
+                }
+            }
+        } else {
+            if let object = self.spriteObject {
+                for variable in UserDataContainer.objectAndProjectLists(for: object) where variable.name == name {
+                    self.askForNewVariableName()
+                    return
+                }
+            }
+        }
+
+        let userList = UserList(name: name)
+
+        if self.newVarIsForProject {
+            self.spriteObject?.scene.project?.userData.add(userList)
+        } else {
+            self.spriteObject?.userData.add(userList)
+        }
+
+        self.spriteObject?.scene.project?.saveToDisk(withNotification: false)
+        self.reloadData()
+    }
+
+    func isVariableUsed(_ userVariable: UserVariable) -> Bool {
+
+        guard let project = self.spriteObject?.scene.project else {
+            fatalError("project of the spriteObject is nil")
+        }
+
+        if project.userData.contains(userVariable) {
+
+            if let objects = self.spriteObject?.scene.objects() {
```

Comment:
Please use `project.allObjects()` instead to get all objects. That's necessary because later we could have several scenes.